internal/idtokenvalidator/idtokenvalidator_test.go

```diff
@@ -0,0 +1,559 @@
+9WD9bRey14VqzZfCvEAYbSWctLeZz9xhz+kP+mWxHvv07wi3/ZiYcVYgOYIhxbnt
+ugNp78oCFUEsdRrdq2juHlqBtjaYpINQSOqNaKUBt3JRkPPxRLyhwCZ6HJVYy5Wp
+4bG6c8eimbglDY4ZwTfUxu1wk3TvMJr9H8tSLwox666NDpxx4ga6kJ06wromShQH
+bCS8WlQKI7CpMbnUSrgNJO6QOvu0N97Szl7UwHFVTEuSvDvMGuUPE+Bn4zbrQVSV
+Go4oqyTchSR7+9xBpzq36OzXM6zSZqmovJuSzMbCXbBjQlo4s61wvnOLhT3D8Z44
+c5VE52frAgMBAAECggEALCx3qXmqNc6AVzDgb+NGfEOT+5dkqQwst0jVoPHswouL
+s998sIoJnngFjwVEFjKZdNrb2i4lb3zlIFzg2qoHurGeoDsQmH7+PNoVs7BL7zm5
+LyLgjsgXt2SB3hoULmtZ9D1byNcG/JrNy6GEDIGuZCSj1T/QPStkwdc+6VpB8pgW
+E8D7jCt40Tik2neYQkDnY775kGAHGWEqpdPCwm+KOnuE1fHx/jk38lmUgYNjKq0h
+JK6Ncjen1X+ZsYfGx4dALWG4cqo3lE0YXXuHuvjJV3aVfzH8t7W4fuZ4+8xvdhhV
+F4br5FimWLbTe2qT4lSpadkbLm3aBlSUR7eAP0BlwQKBgQD5ayZpP5OMp1zfa4hA
+fM8nVUEaVLkRwFK5NChfjHGiaye2RjrnIorXMsFxXjEscgTn2Ux9CgcBhp1fTBhy
+6cmhkp1talAIqLBivNQJT0YTfA+uHrHTTyMfEUgsMzPiiAg7FV7BCG6xd/nsk3yg
+ZUfoXefrhq9LIHsJx7cK12VViQKBgQD4XKvwYmX5t7fZFBPd7dv5ZrcMHQnBMHd7
+is3QhgyKuEgVDzKQ9SA004I9iSvcI3dE/npj31P39N5bbuvYTh4WR/SR4VvXavNG
+AqUR7wm8jTlbiWEPgF9MxC24zaa07Kbxs+P8XT/7wWuijf6+baSFgxQMb80fUArv
+7guKikCo0wKBgCUn3DIDoZRrfj9eQo7wyN9gKPGmO2e0kd47MeSCBI+gjOrvbWjv
+UWWbjwu3b3Xiim6LhYR/EOoeRqViraW4xCvIrqEVHFUd5CDhZmj4oUTXz3It6mnD
+OUUwiuLiwdD2WNuMZHA3NF5FtDqVAhTW4a5xBtKkXsq/TPT5BoCb8+GZAoGAUWAD
+0gpbgTuJ2G10qPWDaq8V8Lke9haMP4VWNCmHuHfy3juRhN9cAxL+DG2CWmmgbZG3
+xjtpRsgLhwfL7J6DyyceYiHltqpLNTgun7ajiQz4qx5TGAImt39bv75aDdOwS2d2
+nrxq93EDdEp0Gi7QhhJRolWLbuQKAV0MmQL9dpMCgYEA5+ug3CDI/jyTHG4ZEVoG
+qmIg7QoHrVEmZrvCMiFw8bbuBvoMnvu1o1zfvAkNrDfibZyxYKHzSqgeVPQShvLa
+P6JCu67ieCGP8C8CMFiQhJ9n4sYGnkzkz67NpkHSzDPA6DfvG4pYuvBQRIefnhYh
+IDGpghhKHMV2DAyzeM4cDU8=
+-----END PRIVATE KEY-----`
+	defaultJWTArgs = jwtArgs{
+		issuer: jwtURL,
+		sub:    "me",
+		exp:    time.Now().Add(3 * time.Second),
+		iat:    time.Now(),
+		aud:    []string{jwtClientID},
+	}
+)
+
+func TestIDTokenValidation(t *testing.T) {
+	t.Run("should error if passed unsupported idTokenSigningAlg", func(t *testing.T) {
+		_, err := New(context.Background(), jwtDomain, jwtClientID, jwtClientSecret, "HS512")
+		assert.ErrorContains(t, err, "Unsupported algorithm")
+	})
+
+	t.Run("validates a token sign by a secret", func(t *testing.T) {
+		args := defaultJWTArgs
+		args.clientSecret = jwtClientSecret
+
+		token, _, err := givenAJWT(t, args)
+		assert.NoError(t, err)
+
+		validator, err := New(context.Background(), jwtDomain, jwtClientID, jwtClientSecret, "HS256")
+		assert.NoError(t, err)
+
+		err = validator.Validate(token, OptionalVerification{})
+		assert.NoError(t, err)
+	})
+
+	t.Run("validates a token sign by a private key", func(t *testing.T) {
+		args := defaultJWTArgs
+		args.privateKey = jwtPrivateKey
+		args.publicKey = jwtPublicKey
+
+		token, server, err := givenAJWT(t, args)
+		assert.NoError(t, err)
+
+		URL, err := url.Parse(server.URL)
+		assert.NoError(t, err)
+
+		validator, err := New(context.Background(), URL.Host, jwtClientID, jwtClientSecret, "RS256", WithHTTPClient(server.Client()))
```

Comment:
Should we mock the JWKS request?

Reply:
Yeah this is all hidden away by the `givenAJWT` helper as it creates the jwt, and then spins up a local HTTPS server that we reference that only serves the created key. I'll change the reference to my tenant to a non-valid string to ensure that no tests are currently piggybacking off that (`validates the signingalg provided is the one used` currently is)